File: AOC.Go/aoc2018/2018_03.go
package aoc2018

import (
	"regexp"
	"strconv"
)

func Solve2018_03_1(lines []string) int64 {
	overlapping, _ := solve(lines)
	return overlapping
}

func Solve2018_03_2(lines []string) int64 {
	_, nonOverlappingId := solve(lines)
	return nonOverlappingId
}

func solve(lines []string) (int64, int64) {
	var grid [1000][1000]int
	allParsed := parse(lines)

	for _, p := range allParsed {
		for row := p.rowOffset; row < p.rowOffset+p.numRows; row++ {
			for col := p.colOffset; col < p.colOffset+p.numCols; col++ {
				grid[row][col]++
			}
		}
	}

	var overlapping int64
	for row := 0; row < 1000; row++ {
		for col := 0; col < 1000; col++ {
			if grid[row][col] > 1 {
				overlapping++
			}
		}
	}

	for i, p := range allParsed {
		isTheOne := true
		for row := p.rowOffset; isTheOne && row < p.rowOffset+p.numRows; row++ {
			for col := p.colOffset; isTheOne && col < p.colOffset+p.numCols; col++ {
				if grid[row][col] > 1 {
					isTheOne = false
				}
			}
		}

		if isTheOne {
			return overlapping, int64(i + 1)
		}
	}

	return -1, -1
}

func parse(lines []string) []Parsed {
	re := regexp.MustCompile(`#\d+ @ (\d+),(\d+): (\d+)x(\d+)`)

	allParsed := make([]Parsed, len(lines))
	for i, line := range lines {
		matches := re.FindStringSubmatch(line)
		colOffset, _ := strconv.Atoi(matches[1])
		rowOffset, _ := strconv.Atoi(matches[2])
		numCols, _ := strconv.Atoi(matches[3])
		numRows, _ := strconv.Atoi(matches[4])

		parsed := Parsed{
			colOffset: colOffset,
			rowOffset: rowOffset,
			numCols:   numCols,
			numRows:   numRows,
		}

		allParsed[i] = parsed
	}

	return allParsed
}

type Parsed struct {
	colOffset int
	rowOffset int
	numCols   int
	numRows   int
}

File: AOC.CSharp/2018/2018_03.cs
using System.Text.RegularExpressions;

namespace AOC.CSharp;

public static class AOC2018_03
{
    private static readonly Regex Re = new(@"#\d+ @ (\d+),(\d+): (\d+)x(\d+)");

    public static long Solve1(string[] lines)
    {
        return Solve(lines).Item1;
    }

    public static long Solve2(string[] lines)
    {
        return Solve(lines).Item2;
    }

    private static (long, long) Solve(string[] lines)
    {
        int[,] grid = new int[1000, 1000];

        List<Parsed> parsedLines = lines.Select(Parse).ToList();
        foreach (Parsed p in parsedLines)
        {
            for (int row = p.RowOffset; row < p.RowOffset + p.NumRows; row++)
            {
                for (int col = p.ColOffset; col < p.ColOffset + p.NumCols; col++)
                {
                    grid[row, col]++;
                }
            }
        }

        long overlapping = 0;
        for (int row = 0; row < 1000; row++)
        {
            for (int col = 0; col < 1000; col++)
            {
                overlapping += grid[row, col] > 1 ? 1 : 0;
            }
        }

        int solution2 = 0;
        foreach (Parsed p in parsedLines)
        {
            bool isTheOne = true;
            for (int row = p.RowOffset; isTheOne && row < p.RowOffset + p.NumRows; row++)
            {
                for (int col = p.ColOffset; isTheOne && col < p.ColOffset + p.NumCols; col++)
                {
                    if (grid[row, col] > 1)
                    {
                        isTheOne = false;
                    }
                }
            }

            solution2++;

            if (isTheOne)
            {
                return (overlapping, solution2);
            }
        }

        return (-1, -1);
    }

    private static Parsed Parse(string line)
    {
        Match m = Re.Match(line);
        return new(
            int.Parse(m.Groups[1].Value),
            int.Parse(m.Groups[2].Value),
            int.Parse(m.Groups[3].Value),
            int.Parse(m.Groups[4].Value)
        );
    }

    private record Parsed(int ColOffset, int RowOffset, int NumCols, int NumRows);
}
